include/core/CStateRestoreTraverser.h

```diff
@@ -59,16 +63,16 @@ class CORE_EXPORT CStateRestoreTraverser : private CNonCopyable {
     //! of this method.
     template<typename FUNC>
     bool traverseSubLevel(FUNC f) {
-        if (!this->hasSubLevel()) {
+        if (this->hasSubLevel() == false || this->haveBadState()) {
             return false;
         }
 
         CAutoLevel level(*this);
```

Comment:
Is the declaration of `level` still needed? (It seems to be no longer used)

Reply:
It's an RAII class, like `scoped_lock`.  Its constructor steps down a level and its destructor will step back up.